### PR TITLE
Implement Pipeline and Processors

### PR DESCRIPTION
**Pull request by p365labs**
This has been a long implementation, Pipelines need to be implemented and even processor if we want to use field attachments.

- Implementation of Pipeline
- Implementation of Processors (only 4 processors are missing : **Foreach, Grok, Gsub, Script**)

This is just a first implementation : I have some doubt on extending **Param** for Processors and for Pipeline Object. It's very useful, 'cause extending ArrayableInterface Object are quickly converted to Arrays.

### CODE REVIEW

**Comment by p365labs** [comment]
I've made some minors update and implemented some tests, in order to execute the pipeline tests against an Index.

File: lib/Elastica/Pipeline.php [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+        return $this->requestEndpoint($endpoint);
+    }
+
+    /**
+     * @todo implement simulate API
+     */
+    public function simulate()
+    {
+
```
Comment (ruflin) [inline lib/Elastica/Pipeline.php]: Should we throw an exception for now?

File: lib/Elastica/Pipeline.php [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+use Elastica\Exception\InvalidException;
+use Elastica\Processor\AbstractProcessor;
+use Elasticsearch\Endpoints\AbstractEndpoint;
+use Elasticsearch\Endpoints\Ingest\Pipeline\Delete;
+use Elasticsearch\Endpoints\Ingest\Pipeline\Get;
+use Elasticsearch\Endpoints\Ingest\Pipeline\Put;
+
+/**
+ * Elastica Processors object.
```
[the remark this answers is not in the document]
Comment (p365labs) [reply lib/Elastica/Pipeline.php]: my fault sorry.

File: lib/Elastica/Pipeline.php (hunk 1 above)
Comment (p365labs) [reply lib/Elastica/Pipeline.php]: ok I raise a *NotImplementedException('simulate API on Pipeline not yet implemented.');*

**Comment by p365labs** [comment]
As for now the community can create and update processor as they change in time, the only processor not yet implemented are **Foreach, Grok, Gsub, Script**.

Param is the right structure to use in order to have the body of the API call well formed, the only concern is that Param was clearly created for queries (at the top of the class --> * * This function can be used to handle params for queries, filter*), and maybe I could create a real similar class called Option or PipelineOption which do exactly the same thing as Param. Maybe is something we can think about in a future if something will change in Pipeline definition or implementation in ES. Creating now an *Option* class would mean only duplicate Param class. I preferred this approach, but I would like to discuss it if it sound ok for you and the community.

Processors does not have functional tests except KvTest.php, I will implement asap, I was waiting for a review first :)

**Comment by ruflin** [comment]
Using Params as you did in this PR LGTM. I initially created it to reduce code duplication in Queries/Filters but I think we can apply it to almost any Class which needs to be transformed from PHP Object to json at one point.

Ping me when you think it's ready to be merged.

**Comment by p365labs** [comment]
Hi @ruflin, before the functional tests I've made a small refactor on *tests*. I created a BasePipeline class (which extends Base), which create a pipeline and also implement a tearDown in order to delete all the pipelines created to make tests atomic. Every **Processor** should extends this class so it would be easy to create a Pipeline (and a functional test...) and also I have avoided to implement the same code to create a Pipeline in all the Processor tests.

**Comment by p365labs** [comment]
Hi @Ruflin the PR for me is ready to be merged; let me know if u have more comments.

**Comment by p365labs** [comment]
@ruflin any comment on the review :) ?

**Comment by ruflin** [comment]
@p365labs Will have a look pretty soon. A bit under water these days ;-)

File: CHANGELOG.md [hunk 3]
```diff
@@ -38,6 +38,7 @@ All notable changes to this project will be documented in this file based on the
 - Added `Query\SpanContaining`, `Query\SpanWithin` and `Query\SpanNot` [#1319](https://github.com/ruflin/Elastica/pull/1319)
 
 ### Improvements
+- Implemented [Pipeline](https://www.elastic.co/guide/en/elasticsearch/reference/current/pipeline.html) and [Processors](https://www.elastic.co/guide/en/elasticsearch/reference/current/ingest-processors.html). [#1373](https://github.com/ruflin/Elastica/pull/1373)
```
Comment (ruflin) [inline CHANGELOG.md]: This should probably go under `Added`

File: lib/Elastica/Client.php [hunk 4]
```diff
@@ -349,6 +356,12 @@ public function addDocuments(array $docs)
 
         $bulk->addDocuments($docs);
 
+        if (!empty($requestParams)) {
```
Comment (ruflin) [inline lib/Elastica/Client.php]: I wonder if the `!empty` is even needed as we already define it as array.

**Comment by p365labs** [comment]
done!

**Pull request merged**